### PR TITLE
Optimize monotonous ORDER BY

### PR DESCRIPTION
**Pull request by 4ertus2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Replaces monotonous functions with its argument in ORDER BY if `set optimize_monotonous_functions_in_order_by=1`.

It's continuation of #11709

### CODE REVIEW

File: src/Interpreters/SyntaxAnalyzer.cpp [hunk 1]
```diff
@@ -529,6 +530,45 @@ void optimizeDuplicateOrderByAndDistinct(ASTPtr & query, const Context & context
...
 
+/// Optimize monotonous functions in ORDER BY
+void optimizeMonotonousFunctionsInOrderBy(ASTSelectQuery * select_query, const Context & context,
+                                          const TablesWithColumns & tables_with_columns)
+{
+    auto order_by = select_query->orderBy();
+    if (!order_by)
+        return;
+
+    std::unordered_map<String, ASTPtr> group_by_hashes;
```
Comment (alesapin) [inline src/Interpreters/SyntaxAnalyzer.cpp:541]: Why unordered_set is not enough? Also we can use `std::unordered_*<UInt128, ASTPtr, UInt128Hash>` directly.

File: src/Interpreters/SyntaxAnalyzer.cpp [hunk 2]
```diff
@@ -529,6 +530,45 @@ void optimizeDuplicateOrderByAndDistinct(ASTPtr & query, const Context & context
     DuplicateDistinctVisitor(distinct_data).visit(query);
 }
 
+/// Optimize monotonous functions in ORDER BY
```
Comment (alesapin) [inline src/Interpreters/SyntaxAnalyzer.cpp:533]: Useless comment. Let's add some implementation details: "Replace monotonous functions in ORDER BY if they don't participate in GROUP BY expression, has a single argument and not an aggregate functions."

File: src/Interpreters/MonotonicityCheckVisitor.h [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            visit(*ast_function, data);
+    }
+
+    static void visit(const ASTFunction & ast_function, Data & data)
+    {
+        if (data.isRejected())
+            return;
+
+        auto arguments = ast_function.arguments;
+        if (arguments->children.size() != 1)
```
Comment (alesapin) [inline src/Interpreters/MonotonicityCheckVisitor.h:87]: Maybe add a comment that it doesn't work for functions with more than one argument.

File: src/Interpreters/MonotonicityCheckVisitor.h [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                return false;
+
+            /// if ORDER BY contains aggregate function it shouldn't be optimized
+            if (AggregateFunctionFactory::instance().isAggregateFunctionName(ast_function.name))
+                return false;
+
+            return true;
+        }
+
+        bool getIdentifierAndType(const ASTFunction & ast_function)
```
Comment (alesapin) [inline src/Interpreters/MonotonicityCheckVisitor.h:53]: setIdentifierAndType?

File: src/Interpreters/MonotonicityCheckVisitor.h [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            identifier = ast_function.arguments->children[0]->as<ASTIdentifier>();
```
Comment (alesapin) [inline src/Interpreters/MonotonicityCheckVisitor.h:55]: It's not clear how it works for nested functions. Maybe add a comment, that we call this function only for the most bottom argument (I can see such comment bellow).

File: src/Interpreters/MonotonicityCheckVisitor.h [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        }
+
+        const auto & function = FunctionFactory::instance().tryGet(ast_function.name, data.context);
+        if (!function)
+        {
+            data.reject();
+            return;
+        }
+
+        /// First time extract the most enclosed identifier and its data type
```
Comment (alesapin) [inline src/Interpreters/MonotonicityCheckVisitor.h:106]: At the deepest nesting level extract the most...? But your comment is also OK.

**Review by alesapin — APPROVED** [review]
LGTM but maybe it worth adding small clarifications.

File: src/Interpreters/MonotonicityCheckVisitor.h (hunk 5 above)
Comment (4ertus2) [reply src/Interpreters/MonotonicityCheckVisitor.h:58]: Noone believes such comments :)
I've add precondition about `identifier` that makes the function more self descriptive.

**Review by alesapin — APPROVED** [review]
LGTM

**Pull request merged**